web/src/components/feature/FluidDialog.tsx

```diff
@@ -22,15 +25,93 @@ const FirstColumn = styled.div`
   gap: 30px;
 `
 
+export type TComponent = {
+  componentId: string
+  altName: string
+  chemicalFormula: string
+  value: number
+}
+
+export type TComponentInput = {
+  [componentId: string]: {
+    altName: string
+    chemicalFormula: string
+    value: number
+  }
+}
+
+function convertComponentResponseToComponentInput(
+  components: ComponentResponse
+): TComponentInput {
+  // convert from ComponentResponse type to TComponentInput
+  return Object.fromEntries(
+    Object.entries(components.components).map(([componentId, names]) => [
+      componentId,
+      { ...names, value: 0 },
+    ])
+  )
+}
+
+function normalizeArray(array: Array<number>): Array<number> {
+  // Normalize array if not "close enough" (|sum - 1| > 0.01) to 1
+  const arraySum: number = array.reduce(
+    (partialSum, value) => partialSum + value,
+    0
+  )
+  return Math.abs(arraySum - 1) > 0.01
+    ? array.map((value) => value / arraySum)
+    : array
+}
+
+function normalizeComponentComposition(
+  componentComposition: TComponentComposition
+): TComponentComposition {
+  // Normalize feedComponentRatios if necessary, else return
+  const normalizedValues: Array<number> = normalizeArray(
+    Object.values(componentComposition)
+  )
+  if (
+    JSON.stringify(normalizedValues) !==
+    JSON.stringify(Object.values(componentComposition))
+  ) {
```

Comment:
A bit overcomplicated. Perhaps just
```suggestion
  const values = Object.values(componentComposition)
  // Normalize feedComponentRatios if necessary, else return
  const normalizedValues: Array<number> = normalizeArray(values)
  if (normalizedValues !== values) {
```

Reply:
Not sure the comparison would work? I think(?) javascript compares arrays as objects meaning that it will return false because the two arrays have different memory addresses.

Reply:
I sent you an image of my console, and then it worked perfectly. But this PR blocks a lot, so I'll just merge it anyway